### PR TITLE
feat: create tickets endpoint

### PR DESCRIPTION
**Pull request by seilagamo**
Create a new endpoint to create tickets. This endpoint calls to vulcan-tracker to create the ticket itself.
Also, the endpoint to retrieve findings has been modified to call vulcan-tracker and get the link of an existing ticket.

The possibility to call vulcan-tracker is decided using a feature flag that will be removed in the feature.

Pending:
- Add testing.

### CODE REVIEW

File: pkg/api/service/vulcantracker.go [hunk 1]
```diff
@@ -0,0 +1,22 @@
+/*
```
Comment (manelmontilla) [inline pkg/api/service/vulcantracker.go:1]: I would name this file, something like: `findings_tickets.go` or just `tickets.go` in order to follow the naming conventions of the package.

File: pkg/common/common.go [hunk 2]
```diff
@@ -33,3 +38,35 @@ func StringValue(v *string) string {
 	}
 	return ""
 }
+
+func IsHttpStatusOk(status int) bool {
```
Comment (manelmontilla) [inline pkg/common/common.go:42]: s/Http/HTTP

Comment (manelmontilla) [reply pkg/common/common.go:42]: In any case, I'm not sure we should create a function in the common package like this. I think that if a HTTP status is okey or not depends a lot on the concrete call to the REST service. For instance,  in this case It's not cleat to me in which cases the vulcan-tracker could return a 300 status code and consider it to be valid.

File: pkg/common/common.go [hunk 3]
```diff
@@ -33,3 +38,35 @@ func StringValue(v *string) string {
...
+	return status >= http.StatusOK && status < http.StatusMultipleChoices
+}
+
+func BuildQueryFilter(filters map[string]string) string {
```
Comment (manelmontilla) [inline pkg/common/common.go:46]: We should add a doc comment in exported functions.

File: pkg/common/common.go [hunk 4]
```diff
@@ -33,3 +38,35 @@ func StringValue(v *string) string {
...
+	filterParts := []string{}
+	for key, value := range filters {
+		part := fmt.Sprintf("%s=%s", key, value)
+		filterParts = append(filterParts, part)
+	}
+	return strings.Join(filterParts, "&")
+}
+
+func ParseHttpErr(statusCode int, mssg string) error {
```
Comment (manelmontilla) [inline pkg/common/common.go:55]: s/Http/HTTP
Also add a doc comment as this is an exported function.

File: pkg/vulcantracker/vulcantracker.go [hunk 5]
```diff
@@ -0,0 +1,137 @@
...
+
+	authScheme = "TEAM team=%s"
+	noAuth     = ""
+)
+
+// Client represents a vulcan tracker client.
+type Client interface {
+	CreateTicket(ctx context.Context, payload api.FindingTicketCreate) (*api.Ticket, error)
+	GetFindingTicket(ctx context.Context, findingID, teamID string) (*api.Ticket, error)
+	IsATeamOnboardedInVulcanTracker(ctx context.Context, teamID string) bool // feature flag.
```
Comment (manelmontilla) [inline pkg/vulcantracker/vulcantracker.go:32]: I would not place this function here. 
As I understand this package it's just a client for the vulcan-tracker component and the feature flag that enables the tracker functionality or not for a component is not independent of that component right?

File: pkg/vulnerabilitydb/vulnerabilitydb.go [hunk 6]
```diff
@@ -513,46 +512,14 @@ func (c *client) performRequest(ctx context.Context, method, path, authTeam stri
...
 			return nil, err
 		}
 
-		return nil, parseErr(resp.StatusCode, string(content))
+		return nil, common.ParseHttpErr(resp.StatusCode, string(content))
 	}
 
 	return ioutil.ReadAll(resp.Body)
 }
-
```
Comment (manelmontilla) [inline pkg/vulnerabilitydb/vulnerabilitydb.go:527]: I wouldn't move these functions: `isStatusOk`, `buildFilter`, `parseErr`  to the package common.
The package common is for more lets say generics things and those functions were only used by the vulnerabilitydb client and now also by the vulcantracker, but not for the rest of the clients, for instance the scanengine. So I would prefer to repeat the code in each of the packages. Also I'm not sure the vulcantracker needs to use those functions right now. I mean the endpoints that the vulcan-tracker is calling don't use query string params and can't return 300 errors right now?

**Review by manelmontilla — CHANGES_REQUESTED** [review]

File: README.md [hunk 7]
```diff
@@ -8,6 +8,8 @@ exposed by Vulcan, the corresponding specifications are here:
 
 ## Requirements
 
+If you are a macOS user you'll need to install `coreutils` before the following commands.
```
[the remark this answers is not in the document]
Comment (seilagamo) [reply README.md:11]: Ok, I have just uninstalled `coreutils` and executed `go generate .\...` and I have got the error:
```
_gen/gen.sh: line 12: realpath: command not found
pkg/asyncapi/vulcan.go:7: running "sh": exit status 127
```

File: pkg/api/endpoint/findings.go [hunk 8]
```diff
@@ -179,6 +179,15 @@ func makeFindFindingEndpoint(s api.VulcanitoService, logger kitlog.Logger) endpo
 			return nil, err
 		}
 
+		findingTicket, err := s.GetFindingTicket(ctx, finding.Finding.ID, r.TeamID)
+		if err != nil {
+			// TODO: we need to decide if we should ignore errors getting tickets from vulcan-tracker
```
[the remark this answers is not in the document]
Comment (seilagamo) [reply pkg/api/endpoint/findings.go:184]: This was when I was thinking about the teams that weren't onboarded in vulcan-tracker. I forgot to refactor this part.

File: README.md (hunk 7 above)
[the remark this answers is not in the document]
Comment (seilagamo) [reply README.md]: I added some extra information about the specific command that produces the error.

Fixed in commit 9267f0cbf7a85be984aed7fdcf524b59d5998404

File: cmd/vulcan-api-cli/design/findings.go [hunk 9]
```diff
@@ -425,3 +440,28 @@ var FindingOverwriteMedia = MediaType("finding_overwrite", func() {
 		Attribute("created_at")
 	})
 })
+
+var FindingTicketCreationMedia = MediaType("finding_ticketcreation", func() {
```
[the remark this answers is not in the document]
Comment (seilagamo) [reply cmd/vulcan-api-cli/design/findings.go]: Fixed in commit 9267f0cbf7a85be984aed7fdcf524b59d5998404

File: cmd/vulcan-api-cli/design/findings.go [hunk 10]
```diff
@@ -425,3 +440,28 @@ var FindingOverwriteMedia = MediaType("finding_overwrite", func() {
...
+	Description("Finding Ticket Creation")
+	Attributes(func() {
+		Attribute("finding_id", String, "Finding ID", func() { Example("3c7d7003-c53d-4ccc-80e7-f21da241b2d4") })
+		Attribute("team_id", String, "The ID associated to the team who requested this ticket creation", func() { Example("a9d33628-ddb0-467c-b482-474887097820") })
+		Attribute("summary", String, "The summary of the ticket", func() { Example("PHP Unsupported Version Detection") })
+		Attribute("description", String, "The previous status for the finding referenced by the finding_id field", func() { Example("Very long desciption explaining the ticket.") })
```
[the remark this answers is not in the document]
Comment (seilagamo) [reply cmd/vulcan-api-cli/design/findings.go]: I have removed all the not necessary fields

Fixed in commit 9267f0cbf7a85be984aed7fdcf524b59d5998404

File: cmd/vulcan-api-cli/design/teams.go [hunk 11]
```diff
@@ -17,12 +17,14 @@ var TeamMedia = MediaType("team", func() {
 		Attribute("name", String, "Name", func() { Example("Security") })
 		Attribute("description", String, "Description", func() { Example("Security Team") })
 		Attribute("tag", String, "tag", func() { Example("team:security") })
+		Attribute("is_onboarded_vulcan_tracker", Boolean, "Is Onboarded in Vulcan Tracker", func() { Example(true) }) // feature flag.
```
[the remark this answers is not in the document]
Comment (seilagamo) [reply cmd/vulcan-api-cli/design/teams.go]: Changed!!

Fixed in commit 9267f0cbf7a85be984aed7fdcf524b59d5998404

File: config.toml [hunk 12]
```diff
@@ -54,6 +54,10 @@ host = "$PERSISTENCE_HOST"
 url = "$VULNERABILITYDB_URL"
 insecure_tls = true
 
+[vulcantracker]
+url = "$VULCANTRACKER_URL"
+insecure_tls = true
+
```
[the remark this answers is not in the document]
Comment (seilagamo) [reply config.toml:62]: Changed in all toml files

Fixed in commit 9267f0cbf7a85be984aed7fdcf524b59d5998404

File: pkg/api/endpoint/findings.go (hunk 8 above)
[the remark this answers is not in the document]
Comment (seilagamo) [reply pkg/api/endpoint/findings.go]: Fixed in commit 9267f0cbf7a85be984aed7fdcf524b59d5998404

File: pkg/api/endpoint/vulcantracker.go [hunk 13]
```diff
@@ -0,0 +1,55 @@
+/*
```
[the remark this answers is not in the document]
Comment (seilagamo) [reply pkg/api/endpoint/vulcantracker.go]: Renamed all vulcantracker words

Fixed in commit 9267f0cbf7a85be984aed7fdcf524b59d5998404

File: pkg/api/endpoint/vulcantracker.go [hunk 14]
```diff
@@ -0,0 +1,55 @@
...
+
+	"github.com/go-kit/kit/endpoint"
+	kitlog "github.com/go-kit/kit/log"
+
+	"github.com/adevinta/errors"
+	"github.com/adevinta/vulcan-api/pkg/api"
+)
+
+// FindingCreateTicketRequest represents a request to Vulcan Tracker that create a relationship
+// between findings and tickets.
```
[the remark this answers is not in the document]
Comment (seilagamo) [reply pkg/api/endpoint/vulcantracker.go]: Fixed in commit 9267f0cbf7a85be984aed7fdcf524b59d5998404

File: pkg/api/service/vulcantracker.go (hunk 1 above)
Comment (seilagamo) [reply pkg/api/service/vulcantracker.go]: Renamed!!

Fixed in commit 9267f0cbf7a85be984aed7fdcf524b59d5998404

File: pkg/common/common.go (hunk 2 above)
Comment (seilagamo) [reply pkg/common/common.go]: I moved the functions to tickets.go and I left vulnerabilitydb.go as it was

Comment (seilagamo) [reply pkg/common/common.go]: Fixed in commit 9267f0cbf7a85be984aed7fdcf524b59d5998404

File: pkg/common/common.go (hunk 3 above)
Comment (seilagamo) [reply pkg/common/common.go]: Added doc comments in all exported functions.
Fixed in commit 9267f0cbf7a85be984aed7fdcf524b59d5998404

File: pkg/vulnerabilitydb/vulnerabilitydb.go (hunk 6 above)
Comment (seilagamo) [reply pkg/vulnerabilitydb/vulnerabilitydb.go]: I left the file as it was before the PR.

Fixed in commit 9267f0cbf7a85be984aed7fdcf524b59d5998404

**Review by manelmontilla — CHANGES_REQUESTED** [review]

File: config.toml [hunk 15]
```diff
@@ -54,6 +54,12 @@ host = "$PERSISTENCE_HOST"
 url = "$VULNERABILITYDB_URL"
 insecure_tls = true
 
+[vulcantracker]
+url = "$VULCANTRACKER_URL"
+insecure_tls = true
+onboarded_teams = "$VULCANTRACKER_TEAMS"
```
[the remark this answers is not in the document]
Comment (seilagamo) [reply config.toml:61]: Fixed in commit 5bed669a367ecd1d0e9c80aba60c5bc3480cef4b

**Review by manelmontilla — DISMISSED** [review]
LGTM

File: README.md [hunk 16]
```diff
@@ -88,6 +91,7 @@ Those are the variables you have to setup:
 |REPORTS_API_URL||http://localhost:8084|
 |PERSISTENCE_HOST||persistence.vulcan.example.com|
 |VULNERABILITYDB_URL||http://localhost:8083|
+|VULCANTRACKER_URL|Leave the url empty if you don't want to configurate the vulcan-tracker component|http://localhost:8085|
```
Comment (manelmontilla) [inline README.md:94]: s/configurate/configure

File: _resources/config/local.toml.example [hunk 17]
```diff
@@ -51,6 +51,13 @@ host = "persistence.vulcan.example.com"
 url = "http://localhost:8083"
 insecure_tls = true
 
+[vulcantracker]
+# leave the url empty if you don't want to configurate the vulcan-tracker component.
```
Comment (manelmontilla) [inline _resources/config/local.toml.example:55]: s/configurate/configure

File: contrib/dev/local-dev.toml [hunk 18]
```diff
@@ -51,6 +51,13 @@ host = "persistence.vulcan.example.com"
 url = "http://localhost:8083"
 insecure_tls = true
 
+[vulcantracker]
+# leave the url empty if you don't want to configurate the vulcan-tracker component.
```
Comment (manelmontilla) [inline contrib/dev/local-dev.toml:55]: s/configurate/configure

File: pkg/api/endpoint/tickets.go [hunk 19]
```diff
@@ -0,0 +1,60 @@
...
+			Summary:     r.Summary,
+			Description: r.Description,
+		}
+
+		if authorizedFindFindingRequest(finding.Finding.Target.Teams, r.TeamID) {
+			ticket, err := s.CreateFindingTicket(ctx, ticketCreate)
+			if err != nil {
+				return nil, err
+			}
+			if ticket == nil {
```
Comment (manelmontilla) [inline pkg/api/endpoint/tickets.go:52]: I wouldn't check here if the ticket is nil and return a 500.
I would modify the method [MakeEndpoints](https://github.com/adevinta/vulcan-api/blob/ddd822ad4ec93b3f640a381da3c144860f1fd92c/pkg/api/endpoint/endpoints.go#L135) so it receives a bool param `isJiraIntegrationEnabled`, and if it is false, the method will not add the `FindingTicketEndpoint`.

File: pkg/api/service/teams.go [hunk 20]
```diff
@@ -13,6 +13,16 @@ import (
 	"github.com/adevinta/vulcan-api/pkg/api"
 )
 
+// isATeamOnboardedInVulcanTracker return if a team is onboarded in vulcan tracker.
+func isATeamOnboardedInVulcanTracker(teamID string, onboardedTeams []string) bool {
```
Comment (manelmontilla) [inline pkg/api/service/teams.go:17]: I would make this function to be a method of the `vulcanitoServiceStruct` and change the implementation to something like this:
```go
if s.vulcantrackerClient == nil {
		return false
	}
	for _, team := range onboardedTeams {
		if team == teamID {
			return true
		}
	}
	return false
``` 
This way we won't need the check if the `vulcanTrackerClient` is nil before calling this function.

File: pkg/api/service/teams.go [hunk 21]
```diff
@@ -39,7 +49,16 @@ func (s vulcanitoService) FindTeam(ctx context.Context, id string) (*api.Team, e
 	if id == "" {
 		return nil, errors.Validation(`ID is empty`)
 	}
-	return s.db.FindTeam(id)
+	team, err := s.db.FindTeam(id)
+	if err != nil {
+		return team, err
+	}
+	if s.vulcantrackerClient == nil {
```
Comment (manelmontilla) [inline pkg/api/service/teams.go:56]: If we do what I suggest in a previous comment we don't need to check if this field is nil here.

File: pkg/api/service/teams.go [hunk 22]
```diff
@@ -53,7 +72,16 @@ func (s vulcanitoService) FindTeamByTag(ctx context.Context, tag string) (*api.T
 	if tag == "" {
 		return nil, errors.Validation(`Tag is empty`)
 	}
-	return s.db.FindTeamByTag(tag)
+	team, err := s.db.FindTeamByTag(tag)
+	if err != nil {
+		return team, err
+	}
+	if s.vulcantrackerClient == nil {
+		return team, nil
```
Comment (manelmontilla) [inline pkg/api/service/teams.go:80]: If we do what I suggest in a previous comment we don't need to check if this field is nil here.

File: pkg/api/service/teams.go [hunk 23]
```diff
@@ -78,5 +106,15 @@ func (s vulcanitoService) DeleteTeam(ctx context.Context, id string) error {
 }
 
 func (s vulcanitoService) ListTeams(ctx context.Context) ([]*api.Team, error) {
-	return s.db.ListTeams()
+	teams, err := s.db.ListTeams()
+	if err != nil {
+		return teams, err
+	}
+	if s.vulcantrackerClient == nil {
+		return teams, nil
```
Comment (manelmontilla) [inline pkg/api/service/teams.go:114]: If we do what I suggest in a previous comment we don't need to check if this field is nil here.

File: pkg/api/service/tickets.go [hunk 24]
```diff
@@ -0,0 +1,27 @@
...
+
+import (
+	"context"
+	"github.com/adevinta/vulcan-api/pkg/api"
+)
+
+// CreateFindingTicket requests the creation of a ticket in the ticket tracker
+// with the values stored in the argument ticket.
+func (s vulcanitoService) CreateFindingTicket(ctx context.Context, ticket api.FindingTicketCreate) (*api.Ticket, error) {
+	if s.vulcantrackerClient == nil {
```
Comment (manelmontilla) [inline pkg/api/service/tickets.go:15]: If don't add ``CreateTicketEndpoint`` if the Jira integration is not enabled we don't need to check if this field is nil here.

**Review by manelmontilla — CHANGES_REQUESTED** [review]

File: README.md (hunk 16 above)
Comment (seilagamo) [reply README.md]: Fixed in commit 3f5b3b3bac78661df03e484824e4d667ea4beadf

File: _resources/config/local.toml.example (hunk 17 above)
Comment (seilagamo) [reply _resources/config/local.toml.example]: Fixed in commit 3f5b3b3bac78661df03e484824e4d667ea4beadf

File: pkg/api/endpoint/tickets.go (hunk 19 above)
Comment (seilagamo) [reply pkg/api/endpoint/tickets.go:52]: Fixed in commit 519b2dce7565f3dd38e8589ab0b487ccb51c4c8c

File: pkg/api/service/teams.go (hunk 20 above)
Comment (seilagamo) [reply pkg/api/service/teams.go]: Fixed in commit 519b2dce7565f3dd38e8589ab0b487ccb51c4c8c

File: pkg/api/service/teams.go (hunk 21 above)
Comment (seilagamo) [reply pkg/api/service/teams.go]: Fixed in commit 519b2dce7565f3dd38e8589ab0b487ccb51c4c8c

File: pkg/api/service/teams.go (hunk 23 above)
Comment (seilagamo) [reply pkg/api/service/teams.go]: Fixed in commit 519b2dce7565f3dd38e8589ab0b487ccb51c4c8c

File: pkg/api/service/teams.go (hunk 22 above)
Comment (seilagamo) [reply pkg/api/service/teams.go]: Fixed in commit 519b2dce7565f3dd38e8589ab0b487ccb51c4c8c

File: pkg/api/service/tickets.go (hunk 24 above)
Comment (seilagamo) [reply pkg/api/service/tickets.go]: Fixed in commit 519b2dce7565f3dd38e8589ab0b487ccb51c4c8c

**Review by manelmontilla — APPROVED** [review]
LGTM

**Pull request merged**